File: 01_basic/week4/49_group-anagrams_watermelon.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const group = {};
    
    for (string of strs) {
        let key = [...string].sort();
        key = key.join("");
        
        if (key in group) {
            group[key] = [...group[key], string];
        } else {
            group[key] = [string];
        }
    }
    
    return [...Object.values(group)];
};

File: 01_basic/week4/49_group-anagrams_watermelon.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        group = {}  # 단어를 사전순 정렬한 값을 키, 원본 단어를 값으로 갖는 딕셔너리
        
        for string in strs:
            key = "".join(sorted(string))  # 단어를 사전순 정렬한 키
            
            # 키가 딕셔너리에 존재하는 경우 원본 단어를 값으로 추가
            if key in group:  
                group[key].append(string)
            
            # 키가 딕셔너리에 존재하지 않는 경우 새로운 키 설정
            else:
                group[key] = [string]
        
        return [value for value in group.values()]
